# Desafio - Cognitivo.AI - `Classificação` **Airbnb**  

- a. Como foi a definição da sua estratégia de modelagem?   
    - Optei pela `classificação` da variável dependente `room_type`  
    - Examinei o conjunto de dados, variáveis e tipagem  
    - Separei as `features` por `tipo`  
        - posso por conveniência juntar estas listas  
    - Escolhi trilhar um caminho de `eliminação de variáveis` e ao longo fui aplicado técnicas estatísticas para este propósito  
        - Modelos com menos número de recursos têm maior explicabilidade  
        - É mais fácil implementar modelos de aprendizado de máquina com um menor número possível de features  
        - Menos features levam a uma generalização aprimorada que, por sua vez, reduz o over-training  
        - A seleção de features remove a redundância de dados  
        - O tempo de treinamento dos modelos com menos features é significativamente menor  
        - Modelos com menos features são menos propensos a erros  
    - Selecionei métodos de `input value` seguindo o método comumente aceito e utilizado para valores `NaN`  
        - `Mediana` para números reais  
        - `Mediana` com `arredondamento` para números inteiros  
        - O caracter `?` nas variáveis `categóricas`    
        - `Moda` nas variáveis `booleanas`  
    - Pelas variáveis categóricas e suas respectivas frequências e distribuições eu selecionei o método da `amostragem`:  
        - `Estratificada` pela variável dependente `room_type`  
            - O `desbalancemaneto` entre as `4 classes` é `alto`   
        - Transformei as `categorias` em `números ordinais`, pois assim possibilitará a utilização de diversos algoritmos baseados em cálculos e estimativas  
        - Apliquei `heuríticas` de `resample`   
            - Afim de avaliar o método que melhor `maximiza` `F1-Score` e `Balanced Accuracy`    
    - Apliquei `heurísticas` de eliminação de `outilers` com o intuito de diminuir `discrepância` entre valores máximos e mínimos e, com isso, de `melhorar` a convergência/ajuste dos algoritmos de classificação tendo cuidado para evitar o `under` e o `over training`  
        - Obs.: no final foi decidido `não` utilizar heurística de eliminação de outliers, pois o resultado sem este uso apresentou ótimos resultados  
            - Desse modo é preferível desconsiderar a remoção dos outiliers  
    - Selecionei a `função de custo`  
        - Neste momento eu aplico a função de custo no dataset original, ou seja, sem minimizar o desbalanceamento   
            - O intuito é `comparar` F1-Score e Balanced Accuracy da `função de custo` com os resultados da classificação destas métricas que utilizaram os datasets com as `heurísticas de desbalanceamento`.  

- b. Como foi definida a função de custo utilizada?   
    - Escolhi a Regressão Logística como função de custo  
        - Por ser um algoritmo de amplo uso e com aplicação em diversas áreas     
    - Utilizei a library `LazyPredict`   
        - Apliquei as heurísticas de desbalanceamento e   
        - E de uma só vez comparei vários algoritmos de classificação com parâmetros standard do LazyPredict   
            - Que inclui também o método utilizado na função de custo, só que desta vez sobre o dataset rebalanceado    
                - Com isso, foi possível a comparação com a função de custo sobre o dataset `original` e nos demais datasets     

- c. Qual foi o critério utilizado na seleção do modelo final?   
    - Maximização do F1-Score e  
    - Maximização do Balanced Accuracy  

- d. Qual foi o critério utilizado para validação do modelo? Por que escolheu utilizar este método?  
    - sklearn.model_selection.train_test_split  
        - Foi feito um cálculo de tamanho amostral, onde foram separados `X_train, ỳ_train` e `X_remaining, y_remaining`  
            - Onde, `X_remaining, y_remaining``X_remaining, y_remaining` foram utilizados no final como conjunto de `teste` 
        - Em `X_train, ỳ_train` foi separado `20%` para `X_valid, y_valid`  
            - Sendo estes últimos utilizados para treinamento e validação, respectivamente  
    - from sklearn.metrics import classification_report   
    - from sklearn.metrics import accuracy_score, balanced_accuracy_score   
    - from sklearn.metrics import confusion_matrix   

- e. Quais evidências você possui de que seu modelo é suficientemente bom?  
    - Chi-Square
    - ROC/AUC# Desafio - Cognitivo.AI - `Classificação` **Airbnb**         

- a. Como foi a definição da sua estratégia de modelagem?  
    - Optei pela `classificação` da variável dependente `room_type`
    - Examinei o conjunto de dados, variáveis e tipagem  
    - Separei as `features` por `tipo`  
        - posso por conveniência juntar estas listas  
    - Escolhi trilhar um caminho de `eliminação de variáveis` e ao longo fui aplicado técnicas estatísticas para este propósito  
        - Modelos com menos número de recursos têm maior explicabilidade  
        - É mais fácil implementar modelos de aprendizado de máquina com um menor número possível de features  
        - Menos features levam a uma generalização aprimorada que, por sua vez, reduz o over-training  
        - A seleção de features remove a redundância de dados  
        - O tempo de treinamento dos modelos com menos features é significativamente menor   
        - Modelos com menos features são menos propensos a erros     
    - Selecionei métodos de `input value` seguindo o método comumente aceito e utilizado para valores `NaN`  
        - `Mediana` para números reais    
        - `Mediana` com `arredondamento` para números inteiros    
        - O caracter `?` nas variáveis `categóricas`    
        - `Moda` nas variáveis `booleanas`  
    - Pelas variáveis categóricas e suas respectivas frequências e distribuições eu selecionei o método da `amostragem`:  
        - `Estratificada` pela variável dependente `room_type`  
            - O `desbalancemaneto` entre as `4 classes` é `alto`   
        - Transformei as `categorias` em `números ordinais`, pois assim possibilitará a utilização de diversos algoritmos baseados em cálculos e estimativas   
        - Apliquei `heuríticas` de `resample`  
            - Afim de avaliar o método que melhor `maximiza` `F1-Score` e `Balanced Accuracy`    
    - Apliquei `heurísticas` de eliminação de `outilers` com o intuito de diminuir `discrepância` entre valores máximos e mínimos e, com isso, de `melhorar` a convergência/ajuste dos algoritmos de classificação tendo cuidado para evitar o `under` e o `over training`   
        - Obs.: no final foi decidido `não` utilizar heurística de eliminação de outliers, pois o resultado sem este uso apresentou ótimos resultados  
            - Desse modo é preferível desconsiderar a remoção dos outiliers    
    - Selecionei a `função de custo`  
        - Neste momento eu aplico a função de custo no dataset original, ou seja, sem minimizar o desbalanceamento   
            - O intuito é `comparar` F1-Score e Balanced Accuracy da `função de custo` com os resultados da classificação destas métricas que utilizaram os datasets com as `heurísticas de desbalanceamento`. 
    

- b. Como foi definida a função de custo utilizada?  
    - Escolhi a Regressão Logística como função de custo  
        - Por ser um algoritmo de amplo uso e com aplicação em diversas áreas    
    - Utilizei a library `LazyPredict`  
        - Apliquei as heurísticas de desbalanceamento e   
        - E de uma só vez comparei vários algoritmos de classificação com parâmetros standard do LazyPredict   
            - Que inclui também o método utilizado na função de custo, só que desta vez sobre o dataset rebalanceado   
                - Com isso, foi possível a comparação com a função de custo sobre o dataset `original` e nos demais datasets     

- c. Qual foi o critério utilizado na seleção do modelo final?  
    - Maximização do F1-Score e  
    - Maximização do Balanced Accuracy  

- d. Qual foi o critério utilizado para validação do modelo? Por que escolheu utilizar este método?  
    - Critério: 
        - sklearn.model_selection.train_test_split  
            - Foi feito um cálculo de tamanho amostral, onde foram separados `X_train, ỳ_train` e `X_remaining, y_remaining`  
                - Onde, `X_remaining, y_remaining``X_remaining, y_remaining` foram utilizados no final como conjunto de `teste`  
            - Em `X_train, ỳ_train` foi separado `20%` para `X_valid, y_valid`  
                - Sendo estes últimos utilizados para treinamento e validação, respectivamente  
        - from sklearn.metrics import classification_report  
        - from sklearn.metrics import accuracy_score, balanced_accuracy_score  
        - from sklearn.metrics import confusion_matrix  
    - Foi escolhido este método por ser um procedimento muito usual e mensura muito bem o desempenho de qualquer algoritmo de classificação  

- e. Quais evidências você possui de que seu modelo é suficientemente bom?  
    - No final há um quadro onde são mostrados os resultados da função de custo `Regressão Logística` comparados com mais três abordagens utilizando o mesmo algoritmo
    
                            sklearn - LogisticRegression     Acc    B-Acc    F1 Score <- FUNÇÃO DE CUSTO <- ORIGINAL  
                                                            0.31     0.70        0.39  
          ------------- LazyPredict -------------------- 
          Sem resample              - LogisticRegression	0.89	 0.52	    0.89 <- FUNÇÃO DE CUSTO <- LazyPredict  
          Tomek                     - LogisticRegression	0.91	 0.51	    0.91  
          NeighbourhoodCleaningRule - LogisticRegression	0.99	 0.57	    0.99 <- MELHOR RESULTADO  

## Carregar libraries, dataset e parâmetros de exibição

In [1]:
#!/anaconda3/envs/picpay/bin/python3
import pandas as pd
# %%capture
from pandarallel import pandarallel
import numpy as np
import scipy.stats as stats

import matplotlib.pyplot as plt
import seaborn as sns

sns.set_theme(style="ticks", color_codes=True)
from IPython.display import Image
%matplotlib inline

import multiprocessing
qtd_cores = multiprocessing.cpu_count() - 1
# https://github.com/nalepae/pandarallel
pandarallel.initialize(nb_workers = qtd_cores)

import warnings
warnings.simplefilter("ignore", DeprecationWarning)
warnings.simplefilter("ignore", Warning)
# warnings.filterwarnings('ignore')

INFO: Pandarallel will run on 7 workers.
INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.


In [2]:
import pandas as pd
pd.set_option('display.max_columns', None)

arq = 'listings.csv' # arquivo
dir = '/home/marcello/Documentos/work/codes/airbnb/' # diretorio
sep = ',' # separador
#enc = encode

def load_dataset(path, arq, sep=',', low_memory=False):
    df = pd.read_csv(path+arq, sep=sep, low_memory=low_memory)
    print('%s' %df.columns)
    return df

df = load_dataset(dir, arq, sep=sep)
# df_raw = load_dataset(dir, arq, sep=sep)
# df = df_raw.copy()

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'name', 'description',
       'neighborhood_overview', 'picture_url', 'host_id', 'host_url',
       'host_name', 'host_since', 'host_location', 'host_about',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_thumbnail_url', 'host_picture_url',
       'host_neighbourhood', 'host_listings_count',
       'host_total_listings_count', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
       'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude',
       'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms',
       'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price',
       'minimum_nights', 'maximum_nights', 'minimum_minimum_nights',
       'maximum_minimum_nights', 'minimum_maximum_nights',
       'maximum_maximum_nights', 'minimum_nights_avg_ntm',
       'maximum_nights_avg_ntm', 'calendar_upd

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26615 entries, 0 to 26614
Data columns (total 74 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            26615 non-null  int64  
 1   listing_url                                   26615 non-null  object 
 2   scrape_id                                     26615 non-null  int64  
 3   last_scraped                                  26615 non-null  object 
 4   name                                          26586 non-null  object 
 5   description                                   25160 non-null  object 
 6   neighborhood_overview                         14314 non-null  object 
 7   picture_url                                   26615 non-null  object 
 8   host_id                                       26615 non-null  int64  
 9   host_url                                      26615 non-null 

In [4]:
object_vars_list = list()
float_vars_list = list()
int_vars_list = list()
numeric_vars_list = list()
reject_vars_list = ['id', 'listing_url', 'scrape_id', 'last_scraped', 'picture_url', 'host_id', 
                    'host_url', 'host_thumbnail_url', 'host_thumbnail_url', 'host_picture_url']

for col in df:
    if pd.api.types.is_object_dtype(df[col]):
        object_vars_list.append(col)
    elif pd.api.types.is_float_dtype(df[col]):
        pd.to_numeric(df[col], downcast='float')
        float_vars_list.append(col)
    elif pd.api.types.is_integer_dtype(df[col]):
        int_vars_list.append(col)
        pd.to_numeric(df[col], downcast='integer')

numeric_vars_list = int_vars_list + float_vars_list
features_all = numeric_vars_list + object_vars_list
label = ['room_type']

# https://stackoverflow.com/questions/2104305/finding-elements-not-in-a-list/2104348
features = list(set(features_all) - set(reject_vars_list))
df[features].head()

maximum_nights  review_scores_cleanliness host_neighbourhood  \
0             180                        9.0         Copacabana   
1              90                       10.0            Ipanema   
2              60                        9.0            Ipanema   
3              89                        9.0            Ipanema   
4              15                       10.0         Copacabana   

   calendar_updated  minimum_nights_avg_ntm  accommodates  \
0               NaN                     5.0             5   
1               NaN                     3.0             2   
2               NaN                     7.0             2   
3               NaN                     2.0             2   
4               NaN                     3.0             2   

  neighbourhood_cleansed instant_bookable  neighbourhood_group_cleansed  \
0             Copacabana                t                           NaN   
1                Ipanema                f                           NaN   
2             Copacabana                f                           NaN   
3                Ipanema                f                           NaN   
4             Copacabana                f                           NaN   

   review_scores_rating  minimum_maximum_nights           neighbourhood  \
0                  94.0                  1125.0  Rio de Janeiro, Brazil   
1                  95.0                    90.0  Rio de Janeiro, Brazil   
2                  94.0                    60.0  Rio de Janeiro, Brazil   
3                  94.0                    89.0  Rio de Janeiro, Brazil   
4                  98.0                     3.0  Rio de Janeiro, Brazil   

   maximum_maximum_nights calendar_last_scraped  bathrooms  \
0                  1125.0            2021-02-23        NaN   
1                    90.0            2021-02-23        NaN   
2                    60.0            2021-02-23        NaN   
3                    89.0            2021-02-24        NaN   
4                    15.0            2021-02-23        NaN   

   host_total_listings_count       property_type has_availability  \
0                        2.0  Entire condominium                t   
1                        0.0    Entire apartment                t   
2                       11.0    Entire apartment                t   
3                        1.0    Entire apartment                t   
4                        1.0         Entire loft                t   

   review_scores_communication  availability_30  license        room_type  \
0                         10.0                0      NaN  Entire home/apt   
1                         10.0                0      NaN  Entire home/apt   
2                         10.0                0      NaN  Entire home/apt   
3                         10.0                0      NaN  Entire home/apt   
4                         10.0                4      NaN  Entire home/apt   

  last_review  review_scores_value  host_since  maximum_nights_avg_ntm  \
0  2021-02-08                  9.0  2010-01-08                  1125.0   
1  2018-02-14                  9.0  2010-03-26                    90.0   
2  2020-02-15                  9.0  2012-10-03                    60.0   
3  2020-03-15                  9.0  2010-06-27                    89.0   
4  2021-02-10                 10.0  2010-06-27                    15.0   

   minimum_nights  review_scores_accuracy  \
0               5                     9.0   
1               3                    10.0   
2               7                     9.0   
3               2                    10.0   
4               3                    10.0   

   calculated_host_listings_count_private_rooms    price  latitude  \
0                                             0  $221.00 -22.96592   
1                                             0  $307.00 -22.98570   
2                                             0  $160.00 -22.97712   
3                                             0  $273.00 -22.98816   
4             

### Sumarização inicial e breve análise sobre algumas variáveis categóricas
- O intuito é conhecer mais sobre os registros e identificar possíveis variáveis para estratificação amostral
- Abaixo se observa o `desbalanceamento` que existe em `room_type`

In [5]:
df.groupby('room_type').size()

room_type
Entire home/apt    19285
Hotel room            99
Private room        6619
Shared room          612
dtype: int64

#### Observa-se acima que a variável dependente `room_type` está `desbalanceada`
- Portanto, trata-se de um problema de classificação cuja `amostragem` deve ser a `estratificada`
- Serão aplicadas `técnicas de desbalanceamento`

In [6]:
round(pd.crosstab(df['host_has_profile_pic'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type             Entire home/apt  Hotel room  Private room  Shared room
host_has_profile_pic                                                        
f                               70.79        0.00         26.97         2.25
t                               72.45        0.37         24.87         2.30

In [7]:
round(pd.crosstab(df['host_response_time'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type           Entire home/apt  Hotel room  Private room  Shared room
host_response_time                                                        
a few days or more            63.19        0.25         32.52         4.04
within a day                  69.56        0.28         26.74         3.41
within a few hours            77.40        0.45         20.37         1.78
within an hour                80.93        0.60         17.29         1.18

In [8]:
round(pd.crosstab(df['host_is_superhost'],df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type          Entire home/apt  Hotel room  Private room  Shared room
host_is_superhost                                                        
f                            71.06        0.33         26.08         2.53
t                            79.18        0.57         19.09         1.16

In [9]:
round(pd.crosstab(df['host_has_profile_pic'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type             Entire home/apt  Hotel room  Private room  Shared room
host_has_profile_pic                                                        
f                               70.79        0.00         26.97         2.25
t                               72.45        0.37         24.87         2.30

In [10]:
round(pd.crosstab(df['host_identity_verified'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type               Entire home/apt  Hotel room  Private room  Shared room
host_identity_verified                                                        
f                                 66.14        0.14         30.20         3.52
t                                 76.15        0.51         21.76         1.58

In [11]:
round(pd.crosstab(df['instant_bookable'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type         Entire home/apt  Hotel room  Private room  Shared room
instant_bookable                                                        
f                           72.87        0.19         24.87         2.07
t                           71.68        0.72         24.87         2.73

In [12]:
round(pd.crosstab(df['has_availability'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type         Entire home/apt  Hotel room  Private room  Shared room
has_availability                                                        
f                           45.22        0.38         48.66         5.73
t                           73.29        0.37         24.15         2.20

In [13]:
round(pd.crosstab(df['calculated_host_listings_count_shared_rooms'], df['room_type']).apply(lambda p: p/p.sum(), axis=1)*100, 2)

room_type                                    Entire home/apt  Hotel room  \
calculated_host_listings_count_shared_rooms                                
0                                                      74.71        0.36   
1                                                      16.35        1.32   
2                                                      13.70        0.00   
3                                                       3.92        0.00   
4                                                       4.05        0.00   
5                                                       0.00        0.00   
6                                                       4.41        0.00   
7                                                       0.00        0.00   
8                                                       0.00        0.00   
9                                                       0.00        0.00   
11                                                      0.00        0.00   

room_type                                    Private room  Shared room  
calculated_host_listings_count_shared_rooms                             
0                                                   24.93         0.00  
1                                                   21.99        60.34  
2                                                   24.66        61.64  
3                                                   25.49        70.59  
4                                                   36.49        59.46  
5                                                    9.09        90.91  
6                                                   33.82        61.76  
7                                                    6.67        93.33  
8                                                   11.11        88.89  
9                                                    5.26        94.74  
11                                                   0.00       100.00

In [14]:
round(pd.crosstab(df['review_scores_accuracy'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type               Entire home/apt  Hotel room  Private room  Shared room
review_scores_accuracy                                                        
2.0                               58.02        0.00         37.04         4.94
3.0                               66.67        0.00         33.33         0.00
4.0                               70.21        0.00         25.53         4.26
5.0                               84.21        0.00         10.53         5.26
6.0                               74.52        0.00         20.38         5.10
7.0                               70.23        0.76         25.95         3.05
8.0                               69.63        0.72         26.07         3.58
9.0                               78.40        0.75         19.55         1.30
10.0                              74.89        0.39         23.25         1.47

In [15]:
round(pd.crosstab(df['review_scores_location'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type               Entire home/apt  Hotel room  Private room  Shared room
review_scores_location                                                        
2.0                               54.35        0.00         36.96         8.70
3.0                              100.00        0.00          0.00         0.00
4.0                               50.00        0.00         40.00        10.00
5.0                               57.14       14.29         14.29        14.29
6.0                               60.53        0.00         35.09         4.39
7.0                               61.90        0.00         31.75         6.35
8.0                               66.00        1.01         28.57         4.43
9.0                               72.49        0.71         24.81         2.00
10.0                              76.00        0.40         22.21         1.38

In [16]:
round(pd.crosstab(df['review_scores_cleanliness'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type                  Entire home/apt  Hotel room  Private room  \
review_scores_cleanliness                                              
2.0                                  65.18        0.00         30.36   
3.0                                  83.33        0.00         16.67   
4.0                                  75.95        0.00         15.19   
5.0                                  75.61        0.00         24.39   
6.0                                  72.89        0.37         24.54   
7.0                                  74.90        0.00         21.62   
8.0                                  74.38        0.40         22.40   
9.0                                  80.26        0.68         18.02   
10.0                                 73.36        0.40         24.71   

room_type                  Shared room  
review_scores_cleanliness               
2.0                               4.46  
3.0                               0.00  
4.0                               8.86  
5.0                               0.00  
6.0                               2.20  
7.0                               3.47  
8.0                               2.82  
9.0                               1.05  
10.0                              1.54

In [17]:
round(pd.crosstab(df['review_scores_checkin'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type              Entire home/apt  Hotel room  Private room  Shared room
review_scores_checkin                                                        
2.0                              58.46        0.00         33.85         7.69
3.0                             100.00        0.00          0.00         0.00
4.0                              73.08        0.00         26.92         0.00
5.0                             100.00        0.00          0.00         0.00
6.0                              66.34        0.00         30.69         2.97
7.0                              69.51        0.00         26.83         3.66
8.0                              75.45        0.91         21.21         2.42
9.0                              77.11        0.74         20.41         1.74
10.0                             75.03        0.43         23.00         1.54

In [18]:
round(pd.crosstab(df['review_scores_communication'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type                    Entire home/apt  Hotel room  Private room  \
review_scores_communication                                              
2.0                                    60.53        0.00         35.53   
3.0                                   100.00        0.00          0.00   
4.0                                    55.17        0.00         41.38   
5.0                                    86.67        0.00         13.33   
6.0                                    76.79        0.00         20.54   
7.0                                    76.40        2.25         15.73   
8.0                                    75.52        0.26         21.13   
9.0                                    77.39        1.11         20.46   
10.0                                   74.87        0.38         23.15   

room_type                    Shared room  
review_scores_communication               
2.0                                 3.95  
3.0                                 0.00  
4.0                                 3.45  
5.0                                 0.00  
6.0                                 2.68  
7.0                                 5.62  
8.0                                 3.09  
9.0                                 1.04  
10.0                                1.59

In [19]:
round(pd.crosstab(df['review_scores_value'], df['room_type']).apply(lambda p: p/p.sum(), axis=1) * 100, 2)

room_type            Entire home/apt  Hotel room  Private room  Shared room
review_scores_value                                                        
2.0                            64.71        0.00         30.59         4.71
3.0                           100.00        0.00          0.00         0.00
4.0                            66.67        0.00         29.41         3.92
5.0                            82.76        0.00         17.24         0.00
6.0                            74.05        0.00         21.11         4.84
7.0                            78.10        0.95         18.57         2.38
8.0                            79.87        0.21         18.45         1.46
9.0                            83.62        0.61         15.11         0.65
10.0                           68.72        0.40         28.76         2.11

### Mostra a frequência unívoca dos valores de cada variável
- Se a frequência for `1` esta variável será `retirada` da lista de `features` e não será considerada na análise

In [20]:
def count_frequence_values(df):
    for col in df:
        if pd.api.types.is_float_dtype(df[col]): 
            resp = len(df[col].unique())
            if resp == 1:
                features.remove(col)
                reject_vars_list.append(col)
            else:
                print('float   -> {} = {}'.format(col, resp))
        elif pd.api.types.is_integer_dtype(df[col]): 
            resp = len(df[col].unique())
            if resp == 1:
                features.remove(col)
                reject_vars_list.append(col)
            else:
                print('integer -> {} = {}'.format(col, resp) )
        elif pd.api.types.is_object_dtype(df[col]): 
            resp = len(df[col].unique())
            if resp == 1:
                features.remove(col)
                reject_vars_list.append(col)
            else:
                print('object  -> {} = {}'.format(col, resp) )
        elif pd.api.types.is_bool_dtype(df[col]): 
            resp = len(df[col].unique())
            if resp == 1:
                features.remove(col)
                reject_vars_list.append(col)
            else:
                print('boolean -> {} = {}'.format(col, resp) )
        else:
            resp = len(df[col].unique())
            if resp == 1:
                features.remove(col)
                reject_vars_list.append(col)
            else:
                print('outros  -> {} = {}'.format(col, resp) )
    return  resp 

count_frequence_values(df[features])
print('\n --- Lista de variáveis rejeitadas ---\n', reject_vars_list)

integer -> maximum_nights = 222
float   -> review_scores_cleanliness = 10
object  -> host_neighbourhood = 157
float   -> minimum_nights_avg_ntm = 180
integer -> accommodates = 17
object  -> neighbourhood_cleansed = 151
object  -> instant_bookable = 2
float   -> review_scores_rating = 49
float   -> minimum_maximum_nights = 190
object  -> neighbourhood = 272
float   -> maximum_maximum_nights = 192
object  -> calendar_last_scraped = 4
float   -> host_total_listings_count = 73
object  -> property_type = 87
object  -> has_availability = 2
float   -> review_scores_communication = 10
integer -> availability_30 = 31
object  -> room_type = 4
object  -> last_review = 1470
float   -> review_scores_value = 10
object  -> host_since = 3393
float   -> maximum_nights_avg_ntm = 351
integer -> minimum_nights = 67
float   -> review_scores_accuracy = 10
integer -> calculated_host_listings_count_private_rooms = 16
object  -> price = 1694
float   -> latitude = 9873
object  -> host_response_rate = 80
float  

## `Exploration Data Analysis` e `Feature Engineering`

### Análise Descritiva

In [21]:
# from pandas_profiling import ProfileReport
# ead = ProfileReport(X_train_transformed[features])
# ead.to_file('ead.html')
# ead.to_notebook_iframe()

### Elimina linhas duplicadas
- É importante considerar a deduplicação **antes da amostragem** para evitar o enviesamento e a distorção dos resultados

In [22]:
tot = df['id'].count()
df_dedup = df.drop_duplicates()
qtd_dedupl = df_dedup['id'].count()
if qtd_dedupl == tot:
    print('Sem duplicidade!')
    print('População: %d - Variáveis: %d' % df.shape)
else:
    dupl = tot - qtd_dedupl
    tot  = tot - dupl
    df.drop_duplicates(inplace=True)
    print('Duplicado: %i' % dupl)
    print('Deduplicado: %i' % tot)
    print('População: %d - Variáveis: %d' % df.shape)

Sem duplicidade!
População: 26615 - Variáveis: 74


#### Outra técnica para encontrar e eliminar a duplicidade  
    - Referência: https://stackabuse.com/applying-filter-methods-in-python-for-feature-selection/

In [23]:
df_T = df.T
print('Duplicidade:', df_T.duplicated().sum())
unique_features_df = df_T.drop_duplicates(keep='first').T
dupl_features = [col for col in df.columns
                 if col not in unique_features_df.columns]
print('Variáveis duplicadas:', dupl_features)
df = unique_features_df
df.shape

Duplicidade: 5
Variáveis duplicadas: ['host_total_listings_count', 'bathrooms', 'calendar_updated', 'calendar_last_scraped', 'license']


(26615, 69)

In [24]:
print('Qtd de features antes:', len(features))
features = [col for col in features
        if col not in dupl_features]
print('Qtd de features depois:', len(features))

Qtd de features antes: 61
Qtd de features depois: 59


#### Quantidade de nulos por coluna

In [25]:
pd.set_option('display.max_rows', round(((df.shape[0]+1) / 100000)))
pd.set_option('display.max_columns', df.shape[1]+1)

print('------ Percentual de Nulos -------')
print(round(100 * df[features].isna().sum() / len(df[features]), 2))


------ Percentual de Nulos -------
maximum_nights                                  0.00
review_scores_cleanliness                      39.48
host_neighbourhood                             42.35
minimum_nights_avg_ntm                          0.32
accommodates                                    0.00
neighbourhood_cleansed                          0.00
instant_bookable                                0.00
review_scores_rating                           39.44
minimum_maximum_nights                          0.32
neighbourhood                                  46.22
maximum_maximum_nights                          0.32
property_type                                   0.00
has_availability                                0.00
                                               ...  
calculated_host_listings_count                  0.00
reviews_per_month                              37.41
host_listings_count                             0.09
calculated_host_listings_count_shared_rooms     0.00
host_identi

## Input Values
- Substituição de valores `NaN` nas variáveis por:
    - float: `mediana`
    - int: `round(mediana)`
    - category: `?`
- Selecionar o método de `input` adequado conforme a tipagem da variável

#### Processo de `input` de valores
- O resultado da iteração abaixo gera as linhas com a sintaxe das funções e métodos de input desejados, conforme o tipo de cada variável

In [26]:
# for col in df[features]:
#     if pd.api.types.is_float_dtype(df[col]): 
#         df['description'] = df['description'].replace(np.nan, '?')
#         print('df["{}"] = df["{}"].replace(np.nan, df["{}"].median())'.format(col,col,col))
#     elif pd.api.types.is_integer_dtype(df[col]): 
#         print('df["{}"] = df["{}"].replace(np.nan, round(df["{}"].median()))'.format(col,col,col))
#     elif pd.api.types.is_object_dtype(df[col]): 
#         print('df["{}"] = df["{}"].replace(np.nan, "?")'.format(col,col))
#     elif pd.api.types.is_bool_dtype(df[col]): 
#         print('df["{}"] = df["{}"].replace(np.nan, df["{}"].mode())'.format(col,col,col))
#     else:
#         print('df["{}"] = df["{}"].replace(np.nan, 0)'.format(col,col,col))

In [27]:
df["minimum_nights"] = df["minimum_nights"].replace(np.nan, "?")
df["number_of_reviews_ltm"] = df["number_of_reviews_ltm"].replace(np.nan, "?")
df["review_scores_value"] = df["review_scores_value"].replace(np.nan, "?")
df["review_scores_checkin"] = df["review_scores_checkin"].replace(np.nan, "?")
df["review_scores_communication"] = df["review_scores_communication"].replace(np.nan, "?")
df["neighbourhood"] = df["neighbourhood"].replace(np.nan, "?")
df["host_location"] = df["host_location"].replace(np.nan, "?")
df["host_has_profile_pic"] = df["host_has_profile_pic"].replace(np.nan, "?")
df["first_review"] = df["first_review"].replace(np.nan, "?")
df["availability_365"] = df["availability_365"].replace(np.nan, "?")
df["calculated_host_listings_count_shared_rooms"] = df["calculated_host_listings_count_shared_rooms"].replace(np.nan, "?")
df["price"] = df["price"].replace(np.nan, "?")
df["property_type"] = df["property_type"].replace(np.nan, "?")
df["host_listings_count"] = df["host_listings_count"].replace(np.nan, "?")
df["host_acceptance_rate"] = df["host_acceptance_rate"].replace(np.nan, "?")
df["availability_60"] = df["availability_60"].replace(np.nan, "?")
df["availability_30"] = df["availability_30"].replace(np.nan, "?")
df["longitude"] = df["longitude"].replace(np.nan, "?")
df["maximum_maximum_nights"] = df["maximum_maximum_nights"].replace(np.nan, "?")
df["minimum_minimum_nights"] = df["minimum_minimum_nights"].replace(np.nan, "?")
df["host_name"] = df["host_name"].replace(np.nan, "?")
df["maximum_nights_avg_ntm"] = df["maximum_nights_avg_ntm"].replace(np.nan, "?")
df["neighborhood_overview"] = df["neighborhood_overview"].replace(np.nan, "?")
df["host_neighbourhood"] = df["host_neighbourhood"].replace(np.nan, "?")
df["last_review"] = df["last_review"].replace(np.nan, "?")
df["has_availability"] = df["has_availability"].replace(np.nan, "?")
df["review_scores_location"] = df["review_scores_location"].replace(np.nan, "?")
df["number_of_reviews_l30d"] = df["number_of_reviews_l30d"].replace(np.nan, "?")
df["host_is_superhost"] = df["host_is_superhost"].replace(np.nan, "?")
df["host_since"] = df["host_since"].replace(np.nan, "?")
df["minimum_nights_avg_ntm"] = df["minimum_nights_avg_ntm"].replace(np.nan, "?")
df["name"] = df["name"].replace(np.nan, "?")
df["availability_90"] = df["availability_90"].replace(np.nan, "?")
df["host_verifications"] = df["host_verifications"].replace(np.nan, "?")
df["calculated_host_listings_count_private_rooms"] = df["calculated_host_listings_count_private_rooms"].replace(np.nan, "?")
df["host_response_time"] = df["host_response_time"].replace(np.nan, "?")
df["minimum_maximum_nights"] = df["minimum_maximum_nights"].replace(np.nan, "?")
df["latitude"] = df["latitude"].replace(np.nan, "?")
df["host_about"] = df["host_about"].replace(np.nan, "?")
df["reviews_per_month"] = df["reviews_per_month"].replace(np.nan, "?")
df["review_scores_cleanliness"] = df["review_scores_cleanliness"].replace(np.nan, "?")
df["review_scores_rating"] = df["review_scores_rating"].replace(np.nan, "?")
df["neighbourhood_cleansed"] = df["neighbourhood_cleansed"].replace(np.nan, "?")
df["bedrooms"] = df["bedrooms"].replace(np.nan, "?")
df["calculated_host_listings_count"] = df["calculated_host_listings_count"].replace(np.nan, "?")
df["number_of_reviews"] = df["number_of_reviews"].replace(np.nan, "?")
df["instant_bookable"] = df["instant_bookable"].replace(np.nan, "?")
df["description"] = df["description"].replace(np.nan, "?")
df["beds"] = df["beds"].replace(np.nan, "?")
df["bathrooms_text"] = df["bathrooms_text"].replace(np.nan, "?")
df["amenities"] = df["amenities"].replace(np.nan, "?")
df["maximum_minimum_nights"] = df["maximum_minimum_nights"].replace(np.nan, "?")
df["calculated_host_listings_count_entire_homes"] = df["calculated_host_listings_count_entire_homes"].replace(np.nan, "?")
df["review_scores_accuracy"] = df["review_scores_accuracy"].replace(np.nan, "?")
df["accommodates"] = df["accommodates"].replace(np.nan, "?")
df["maximum_nights"] = df["maximum_nights"].replace(np.nan, "?")
df["host_identity_verified"] = df["host_identity_verified"].replace(np.nan, "?")
df["room_type"] = df["room_type"].replace(np.nan, "?")
df["host_response_rate"] = df["host_response_rate"].replace(np.nan, "?")

- O método com as funções abaixo tiveram um péssimo desempenho;
- Eu tentei diferentes formas e métodos. Na documentação diz que o uso de `apply ou applymap` otimiza a função lambda em questão;
- Fiz paralelização e os resultados foram bem piores. A memória RAM e swap (SSD) foram totalmente usadas, sendo o processamento no sentido contrário;
- Por fim, a solução na célula acima funcionou num piscar de olhos.

In [28]:
# def input_value(df):
#     if pd.api.types.is_float_dtype(df):
#         media = df.mean()
#         resp = df.fillna(media)
#     elif pd.api.types.is_integer_dtype(df):
#         moda = df.mode(df)
#         resp = df.fillna(moda)
#     elif pd.api.types.is_object_dtype(df):
#         resp = df.fillna('?')
#     else:
#         resp = 0
#     return resp

In [29]:
# from tqdm import tqdm
# for col in df[features]:
#     tqdm.pandas(desc='%s' %col.strip())
#     proc_col = lambda x: input_value(df[col])
#     parallel_col = lambda x: df.parallel_apply(proc_col)
#     df[col] = df[col].progress_apply(parallel_col)
# df[features]    

#### Sumariza por percentual os valores únicos de cada variável
- Mostra: 
    - coluna 
    - número de valores distintos por coluna
    - porcentagem de valores distintos de todas as linhas no conjunto de dados. 
- Observam-se colunas com porcentagem de valores distintos muito baixa
    - abaixo de 1%.
- Isso não significa que essas linhas e colunas devam ser excluídas, mas requerem mais atenção.
    - É o caso de transformar os valores distintos em valores ordinais
    - Os valores ordinais distintos podem ser considerados como valores categóricos
        - valores numéricos são necessários, pois vários algoritmos/técnicas funcionam, somente, com números
    - Será aplicado o processo de transformar variáveis categóricas em números ordinais

In [30]:
# Percentual de valores distintos por coluna
for i in range(df[features].shape[1]):
    col = df.columns
    num = len(df.iloc[:, i].unique())
    percent = float(num) / df.shape[0] * 100
    print('%s: %d  %.3f%%' % (col[i], num, percent))

id: 26615  100.000%
listing_url: 26615  100.000%
scrape_id: 1  0.004%
last_scraped: 4  0.015%
name: 25807  96.964%
description: 24459  91.899%
neighborhood_overview: 13122  49.303%
picture_url: 26129  98.174%
host_id: 17324  65.091%
host_url: 17324  65.091%
host_name: 5146  19.335%
host_since: 3393  12.748%
host_location: 964  3.622%
host_about: 6976  26.211%
host_response_time: 5  0.019%
host_response_rate: 80  0.301%
host_acceptance_rate: 95  0.357%
host_is_superhost: 3  0.011%
host_thumbnail_url: 17221  64.704%
host_picture_url: 17221  64.704%
host_neighbourhood: 157  0.590%
host_listings_count: 73  0.274%
host_verifications: 358  1.345%
host_has_profile_pic: 3  0.011%
host_identity_verified: 3  0.011%
neighbourhood: 272  1.022%
neighbourhood_cleansed: 151  0.567%
neighbourhood_group_cleansed: 1  0.004%
latitude: 9873  37.096%
longitude: 12257  46.053%
property_type: 87  0.327%
room_type: 4  0.015%
accommodates: 17  0.064%
bathrooms_text: 53  0.199%
bedrooms: 19  0.071%
beds: 37  0.

#### Coloca para `category` as variáveis de tipo `object` e `boolean`

In [31]:
def df_change_type(df, features):
    for atrib in df[features]:
        if pd.api.types.is_object_dtype(df[atrib]):
            df[atrib] = df[atrib].astype('category')
        if pd.api.types.is_bool_dtype(df[atrib]):
            df[atrib] = df[atrib].astype('category') 
    return df

df = df_change_type(df, features)
df.shape

(26615, 69)

### Transformação dos Dados
#### Transformar variáveis `categóricas` em números `ordinais`
- Tipo de variáveis `object` e `bool` mudarão para `category`

In [32]:
#import pandas.util.testing as tm
import category_encoders as ce
encoder = ce.OrdinalEncoder()
y_encoder = ce.OrdinalEncoder()
X_train_transformed = encoder.fit_transform(df[features], df[label])
y_train_transformed = y_encoder.fit_transform(df[label])
print(X_train_transformed.shape)
y_train_transformed[label].groupby(label).size()

(26615, 59)


room_type
1    19285
2     6619
3      612
4       99
dtype: int64

- 1 = Entire home/apt 
- 2 = Private room
- 3 = Shared room
- 4 = Hotel room

# Amostragem para treinamento do modelo
- Baseado no cálculo de amostragem em https://solvis.com.br/calculos-de-amostragem/ (Calculadora 2)
    - Método amostral: estratificada proporcional
        - Obtem as mesmas proporções observadas na população
    - População 26.615
    - Margem de erro (%): 1
    - Confiabilidade (%): 99
    - Tamanho da amostra: 10.240
    - Remanescentes : 16.375

- Referências:
    - LARSON, Ron. Estatística Aplicada. São Paulo. Person Educational do Brasil, 2015.
    - https://en.wikipedia.org/wiki/Stratified_samplinghttps://en.wikipedia.org/wiki/Stratified_sampling
    - https://www.netquest.com/blog/br/blog/br/amostragem-probabilistica-amostra-estratificada

In [33]:
from sklearn.model_selection import train_test_split
features = set(features) - set(label) # retira LABEL (room_type) da lista de features
# X_train, X_valid, y_train, y_valid = train_test_split(X_train_transformed[features].drop(labels=label, axis=1), X_train_transformed[label],
X_train, X_remaining, y_train, y_remaining = train_test_split(X_train_transformed[features], X_train_transformed[label],
                                                              test_size=0.61525,
                                                              random_state=49,
                                                              stratify=X_train_transformed[label])

print((X_train.shape, y_train.shape))
print((X_remaining.shape, y_remaining.shape))

((10240, 58), (10240, 1))
((16375, 58), (16375, 1))


In [34]:
from sklearn.model_selection import train_test_split
# features = set(features) - set(label) # retira LABEL (room_type) da lista de features
X_train, X_valid, y_train, y_valid = train_test_split(X_train[features], y_train[label],
                                                              test_size=0.20,
                                                              random_state=7,
                                                              stratify=y_train[label])

print((X_train.shape, y_train.shape))
print((X_valid.shape, y_valid.shape))

((8192, 58), (8192, 1))
((2048, 58), (2048, 1))


### Selecionar variáveis com alta correlação e baixa informação mútua
- Um limite baseado em variância é uma das primeira etapa de seleção de variáveis (features)
- A variação pode ser usada como um filtro para identificar colunas a serem removidas do conjunto de dados. 
- Uma coluna com um único valor tem uma variação de 0.0, e uma coluna com muito poucos valores exclusivos pode ter uma pequena variação
- Objetivo
    - Reduzir a incerteza de uma variável
    - Correlação
        - Baixa: 0,10 a 0,29
        - Médio: 0,30 a 0,49
        - Alta: 0,50 a 1,00
    - Os intervalos também se aplicam a valores negativos entre 0 e -1
    - Colunas que têm uma variação menor que **`0.50`** serão removidas

- Referências:
    - https://eng.libretexts.org/Bookshelves/Industrial_and_Systems_Engineering/Book%3A_Chemical_Process_Dynamics_and_Controls_(Woolf)/13%3A_Statistics_and_Probability_Background/13.13%3A_Correlation_and_Mutual_Information
    - https://en.wikipedia.org/wiki/Mutual_information
    - https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.VarianceThreshold.html
    - https://stackabuse.com/applying-filter-methods-in-python-for-feature-selection

In [35]:
print(X_train.shape, X_valid.shape)
print(y_train.groupby('room_type').size())

(8192, 58) (2048, 58)
room_type
1    5936
2    2038
3     188
4      30
dtype: int64


In [36]:
from sklearn.feature_selection import VarianceThreshold

constant_filter = VarianceThreshold(threshold=0.55)
constant_filter.fit(X_train)
constant_cols = [col for col in X_train.columns
                if col not in X_train.columns[constant_filter.get_support()]]
print('\n----------- Variáveis removidas -----------\n %s' %constant_cols)

remaining_cols = [col for col in X_train.columns
                 if col in X_train.columns[constant_filter.get_support()]]
print('\nVariáveis restantes com variação acima 50%% -> %d' % len(X_train.columns[constant_filter.get_support()]))
print('\n----------- Variáveis restantes -----------\n %s' %remaining_cols)

X_train = constant_filter.transform(X_train)
X_valid = constant_filter.transform(X_valid)

X_train = pd.DataFrame(data=X_train, columns=remaining_cols)
X_valid = pd.DataFrame(data=X_valid, columns=remaining_cols)

X_remaining = constant_filter.transform(X_remaining)
X_remaining = pd.DataFrame(data=X_remaining, columns=remaining_cols)

X_train.shape, X_valid.shape, X_remaining.shape


----------- Variáveis removidas -----------
 ['latitude', 'instant_bookable', 'host_has_profile_pic', 'has_availability', 'host_is_superhost', 'calculated_host_listings_count_shared_rooms', 'host_identity_verified', 'longitude']

Variáveis restantes com variação acima 50% -> 50

----------- Variáveis restantes -----------
 ['maximum_nights', 'review_scores_cleanliness', 'host_neighbourhood', 'minimum_nights_avg_ntm', 'host_response_rate', 'accommodates', 'neighbourhood_cleansed', 'bedrooms', 'number_of_reviews', 'host_response_time', 'review_scores_rating', 'minimum_maximum_nights', 'review_scores_location', 'maximum_minimum_nights', 'neighbourhood', 'maximum_maximum_nights', 'host_acceptance_rate', 'description', 'availability_365', 'review_scores_checkin', 'availability_60', 'number_of_reviews_ltm', 'first_review', 'property_type', 'number_of_reviews_l30d', 'review_scores_communication', 'availability_30', 'host_about', 'neighborhood_overview', 'host_location', 'bathrooms_text', 'ho

((8192, 50), (2048, 50), (16375, 50))

#### Removendo features `correlacionadas` acima de 80%
- Ref: https://stackabuse.com/applying-filter-methods-in-python-for-feature-selection/

In [37]:
corr_features = set()
corr_matrix = X_train.corr()
    
for i in range(len(corr_matrix.columns)):
    for j in range(i):
        if abs(corr_matrix.iloc[i,j]) > 0.8:
            col = corr_matrix.columns[i]
            corr_features.add(col)
print('Total de variáveis com correlação acima de 80%% -> %d' %len(corr_features))
print('\n---------- Variáveis removidas ----------\n%s' %corr_features)

features = list(X_train.columns[1])
print('\n---------- Variáveis remanescentes ----------\n%s' %features)

X_train.drop(labels=corr_features, axis=1, inplace=True)
X_valid.drop(labels=corr_features, axis=1, inplace=True)
X_remaining.drop(labels=corr_features, axis=1, inplace=True)

X_train.shape, X_valid.shape, X_remaining.shape

Total de variáveis com correlação acima de 80% -> 10

---------- Variáveis removidas ----------
{'minimum_minimum_nights', 'calculated_host_listings_count_entire_homes', 'review_scores_accuracy', 'amenities', 'calculated_host_listings_count_private_rooms', 'review_scores_communication', 'availability_30', 'availability_90', 'maximum_maximum_nights', 'review_scores_value'}


((8192, 40), (2048, 40), (16375, 40))

## Análise de outliers

In [38]:
def diagnostic_number_plots(df, variable):
    for v in variable:
        if pd.api.types.is_numeric_dtype(df[v]):
            # define o tamanho
            plt.figure(figsize=(16, 4))

            # histogram
            plt.subplot(1, 3, 1)
            sns.distplot(df[v], bins=30)
            plt.title('Histogram')

            # Q-Q plot
            plt.subplot(1, 3, 2)
            stats.probplot(df[v], dist="norm", plot=plt)
            plt.ylabel(v+' quantiles')

            # boxplot
            plt.subplot(1, 3, 3)
            sns.boxplot(y=df[v])
            plt.title('Boxplot')

            plt.show()

In [39]:
# diagnostic_number_plots(X_train_transformed, features)
# diagnostic_number_plots(X_train_transformed_variance, features)

### Função para `separar os outliers`. Retorna um dataframe de `outliers` e outro `sem outliers`
- Referências:
    - https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.LocalOutlierFactor.htmlhttps://scikit-learn.org/stable/modules/generated/sklearn.neighbors.LocalOutlierFactor.html
    - https://scikit-learn.org/stable/auto_examples/neighbors/plot_lof_outlier_detection.htmlhttps://scikit-learn.org/stable/auto_examples/neighbors/plot_lof_outlier_detection.html
    - https://towardsdatascience.com/anomaly-detection-with-local-outlier-factor-lof-d91e41df10f2https://towardsdatascience.com/anomaly-detection-with-local-outlier-factor-lof-d91e41df10f2
    - https://towardsdatascience.com/why-1-5-in-iqr-method-of-outlier-detection-5d07fdc82097https://towardsdatascience.com/why-1-5-in-iqr-method-of-outlier-detection-5d07fdc82097

In [40]:
# from numpy import percentile
# def remove_outliers(df, cols, limit=1.5):
#     outliers = pd.DataFrame(columns=cols)
#     df_clean = pd.DataFrame(columns=cols)
    
#     for i, col in enumerate(df):
#         q25, q75 = percentile(df.iloc[i], 25), percentile(df.iloc[i], 75)
#         irq = q75 - q25

#         threshold = irq * limit
#         lower, upper = q25 - threshold, q75 + threshold

#         df_clean[[df.iloc[i].index]] = np.array([x for x in df.iloc[i] if x >= lower or x <= upper])

#     return outliers, df_clean

# outliers, df_clean = remove_outliers(X_train_transformed, [features], 1.5)

# print('Registros: %i | Variáveis atuais: %i' %X_train_transformed[features].shape)
# print('Registros: %i | Variáveis df_clean: %i' %df_clean.shape)
# print('Registros: %i | Variáveis outliers: %i' %outliers.shape)

In [41]:
# https://stackoverflow.com/questions/31323499/sklearn-error-valueerror-input-contains-nan-infinity-or-a-value-too-large-for
def clean_dataset(df):
    assert isinstance(df, pd.DataFrame), "df deve ser um pd.DataFrame"
    df.dropna(inplace=True)
    indices_to_keep = ~df.isin([np.nan, np.inf, -np.inf]).any(1)
    return df[indices_to_keep].astype(np.float64)

X_train = clean_dataset(X_train)
X_valid = clean_dataset(X_valid)

In [42]:
from sklearn.neighbors import LocalOutlierFactor

# atualiza a lista de features
features = list(X_train.columns)

lof = LocalOutlierFactor(n_neighbors=20, contamination=0.15, metric='minkowski', n_jobs=qtd_cores)
outliers = lof.fit_predict(X_train[features])
check = outliers != -1
X_train_clean_lof = X_train[features].iloc[check, :]

print('Atual: %i | Variáveis atuais: %i' %X_train[features].shape)
print('Limpo: %i | Variáveis restantes: %i' %X_train_clean_lof.shape)

Atual: 8192 | Variáveis atuais: 40
Limpo: 6963 | Variáveis restantes: 40


In [43]:
# from sklearn.ensemble import IsolationForest

# iso = IsolationForest(contamination=0.15, n_jobs=qtd_cores)
# outliers = iso.fit_predict(X_train[features])
# check = outliers != -1
# X_train_clean_iso = X_train[features].iloc[check, :]

# print('Atual: %i | Variáveis atuais: %i' %X_train[features].shape)
# print('Limpo: %i | Variáveis restantes: %i' %X_train_clean_iso.shape)

In [44]:
# from sklearn.covariance import EllipticEnvelope

# epe = EllipticEnvelope(contamination=0.15)
# outliers = epe.fit_predict(X_train[features])
# check = outliers != -1
# X_train_clean_epe = X_train[features].iloc[check, :]

# print('Atual: %i | Variáveis atuais: %i' %X_train[features].shape)
# print('Limpo: %i | Variáveis restantes: %i' %X_train_clean_epe.shape)

In [45]:
# from sklearn.svm import OneClassSVM

# one = EllipticEnvelope(contamination=0.15)
# outliers = one.fit_predict(X_train[features])
# check = outliers != -1
# X_train_clean_one = X_train[features].iloc[check, :]

# print('Atual: %i | Variáveis atuais: %i' %X_train[features].shape)
# print('Limpo: %i | Variáveis restantes: %i' %X_train_clean_one.shape)

In [46]:
# diagnostic_number_plots(X_train_clean4, features)

# Modelo de `Regressão Logística` criado com o objetivo de gerar a `função de custo`
- Será usado no dataset desbalanceado
    - Servirá como parâmetro de comparação, pois utilizará o dataset sem desbalanceamento
    - Portanto, os modelos gerados pelo LazyPredict utilizarão as heurísticas undersampling acima preparadas
  - Referência:
      - https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html
      - https://scikit-learn.org/stable/modules/model_evaluation.html

In [47]:
# Modelo Regressão Logística
from sklearn.linear_model import LogisticRegression
# Métricas
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score, balanced_accuracy_score
from sklearn.metrics import confusion_matrix

# O parâmetro class_weight='balanced' permite que o algoritmo considere peso maior
# para a classe minoritária. Busca-se um efeito de ajustamento que melhor
# considere a classe minoritária Sim devido a grande desproporcionalidade.
model_cost = LogisticRegression(class_weight='balanced', n_jobs=-1)
model_cost.fit(X_train, y_train.iloc[:,0])
y_pred_cost = model_cost.predict(X_valid)

label_category = [1,2,3,4]

print('Acurácia: %.2f' % accuracy_score(y_valid.iloc[:,0], y_pred_cost))
print('Acurácia balanceada: %.2f' % balanced_accuracy_score(y_valid.iloc[:,0], y_pred_cost))
print(classification_report(y_valid.iloc[:,0], y_pred_cost, zero_division=0))
print(confusion_matrix(y_valid.iloc[:,0], y_pred_cost))
print('########################################################\n')

Acurácia: 0.28
Acurácia balanceada: 0.37
              precision    recall  f1-score   support

           1       0.84      0.29      0.43      1484
           2       0.36      0.24      0.28       509
           3       0.04      0.45      0.07        47
           4       0.01      0.50      0.01         8

    accuracy                           0.28      2048
   macro avg       0.31      0.37      0.20      2048
weighted avg       0.70      0.28      0.39      2048

[[436 208 376 464]
 [ 81 120 178 130]
 [  4   9  21  13]
 [  0   1   3   4]]
########################################################



## Resample
- Utilização de técnicas visando minimizar o desbalanceamento que existe na variável dependente `room_type`  

- Referências:
    - https://pypi.org/project/imbalanced-learn/
    - https://imbalanced-learn.org/stable/

### Gera NeighbourhoodCleaningRule `undersample`

In [48]:
from imblearn.under_sampling import NeighbourhoodCleaningRule

ncr = NeighbourhoodCleaningRule(n_neighbors=20, threshold_cleaning=0.5, n_jobs=-1)
X_ncr_train, y_ncr_train = ncr.fit_resample(X_train[features], y_train)
X_ncr_valid, y_ncr_valid = ncr.fit_resample(X_valid[features], y_valid)

### Gera TomekLinks `undersample`

In [49]:
from imblearn.under_sampling import TomekLinks

tmk = TomekLinks(n_jobs=-1)
X_tmk_train, y_tmk_train = tmk.fit_resample(X_train, y_train)
X_tmk_valid, y_tmk_valid = tmk.fit_resample(X_valid, y_valid)

### Cria os modelos com `LazyPredict` e aplica `Resamples`
- Referências:
    - https://pypi.org/project/lazypredict/https://pypi.org/project/lazypredict/
    - https://lazypredict.readthedocs.io/en/latest/https://lazypredict.readthedocs.io/en/latest/
    - https://medium.com/swlh/lazy-predict-for-ml-models-c513a5daf792https://medium.com/swlh/lazy-predict-for-ml-models-c513a5daf792

In [50]:
# import pandas.util.testing as tm
from lazypredict.Supervised import LazyClassifier
# from lazypredict.Supervised import LazyRegressor
clf = LazyClassifier(verbose=0, predictions=True, ignore_warnings=True, custom_metric=None)

- Utiliza `NeighbourhoodCleaningRule` resample

In [51]:
pd.set_option('display.max_rows', 50)
pd.set_option('display.max_columns', X_train.shape[1]+1)
clf_ncr = LazyClassifier(verbose=0, ignore_warnings=True, custom_metric=None)
models_class_ncr, predictions_class_ncr = clf_ncr.fit(X_ncr_train, X_ncr_valid, y_ncr_train, y_ncr_valid)
models_class_ncr

100%|██████████| 29/29 [00:27<00:00,  1.06it/s]


Accuracy  Balanced Accuracy ROC AUC  F1 Score  \
Model                                                                          
XGBClassifier                      1.00               0.84    None      1.00   
BaggingClassifier                  1.00               0.84    None      1.00   
DecisionTreeClassifier             0.99               0.84    None      0.99   
NearestCentroid                    0.81               0.83    None      0.89   
LGBMClassifier                     1.00               0.80    None      1.00   
GaussianNB                         0.73               0.76    None      0.83   
LinearDiscriminantAnalysis         0.98               0.69    None      0.98   
Perceptron                         0.99               0.69    None      0.99   
RandomForestClassifier             0.99               0.68    None      0.99   
ExtraTreeClassifier                0.98               0.60    None      0.98   
AdaBoostClassifier                 0.99               0.57    None      0.99   
PassiveAggressiveClassifier        0.99               0.57    None      0.99   
SGDClassifier                      0.99               0.56    None      0.99   
LabelSpreading                     0.99               0.49    None      0.99   
LabelPropagation                   0.99               0.49    None      0.99   
LogisticRegression                 0.99               0.49    None      0.99   
KNeighborsClassifier               0.99               0.46    None      0.99   
LinearSVC                          0.99               0.45    None      0.99   
BernoulliNB                        0.97               0.45    None      0.98   
QuadraticDiscriminantAnalysis      0.94               0.39    None      0.96   
ExtraTreesClassifier               0.99               0.38    None      0.98   
CalibratedClassifierCV             0.99               0.38    None      0.98   
SVC                                0.99               0.37    None      0.98   
RidgeClassifier                    0.99               0.33    None      0.98   
RidgeClassifierCV                  0.99               0.33    None      0.98   
DummyClassifier                    0.97               0.33    None      0.97   

                               Time Taken  
Model                                      
XGBClassifier                        1.32  
BaggingClassifier                    0.24  
DecisionTreeClassifier               0.06  
NearestCentroid                      0.07  
LGBMClassifier                       1.55  
GaussianNB                           0.06  
LinearDiscriminantAnalysis           0.19  
Perceptron                           0.07  
RandomForestClassifier               1.14  
ExtraTreeClassifier                  0.05  
AdaBoostClassifier                   1.13  
PassiveAggressiveClassifier          0.09  
SGDClassifier                        0.28  
LabelSpreading                       7.74  
LabelPropagation                     5.83  
LogisticRegression                   1.15  
KNeighborsClassifier                 1.38  
LinearSVC                            0.68  
BernoulliNB                          0.08  
QuadraticDiscriminantAnalysis        0.11  
ExtraTreesClassifier                 0.78  
CalibratedClassifierCV               2.54  
SVC                                  0.42  
RidgeClassifier                      0.09  
RidgeClassifierCV                    0.18  
DummyClassifier                      0.04

- Utiliza `TomekLink` resample

In [52]:
clf_tmk = LazyClassifier(verbose=0, predictions=True, ignore_warnings=True, custom_metric=None)
models_class_tmk, predictions_class_tmk = clf_tmk.fit(X_tmk_train, X_tmk_valid, y_tmk_train, y_tmk_valid)
models_class_tmk

100%|██████████| 29/29 [00:51<00:00,  1.78s/it]


Accuracy  Balanced Accuracy ROC AUC  F1 Score  \
Model                                                                          
DecisionTreeClassifier             0.99               0.96    None      0.99   
LGBMClassifier                     0.99               0.87    None      0.99   
XGBClassifier                      0.99               0.87    None      0.99   
BaggingClassifier                  0.99               0.84    None      0.99   
NearestCentroid                    0.76               0.78    None      0.79   
RandomForestClassifier             0.98               0.63    None      0.98   
BernoulliNB                        0.93               0.62    None      0.93   
LinearDiscriminantAnalysis         0.89               0.60    None      0.88   
ExtraTreesClassifier               0.96               0.58    None      0.96   
GaussianNB                         0.54               0.55    None      0.63   
ExtraTreeClassifier                0.88               0.50    None      0.88   
AdaBoostClassifier                 0.97               0.49    None      0.96   
PassiveAggressiveClassifier        0.89               0.47    None      0.88   
LogisticRegression                 0.91               0.47    None      0.90   
SGDClassifier                      0.90               0.47    None      0.89   
SVC                                0.93               0.46    None      0.92   
Perceptron                         0.87               0.46    None      0.86   
CalibratedClassifierCV             0.91               0.45    None      0.90   
LinearSVC                          0.90               0.45    None      0.89   
QuadraticDiscriminantAnalysis      0.58               0.44    None      0.66   
LabelSpreading                     0.85               0.43    None      0.84   
LabelPropagation                   0.85               0.43    None      0.84   
RidgeClassifier                    0.89               0.40    None      0.88   
RidgeClassifierCV                  0.89               0.40    None      0.88   
KNeighborsClassifier               0.86               0.40    None      0.85   
DummyClassifier                    0.66               0.26    None      0.66   

                               Time Taken  
Model                                      
DecisionTreeClassifier               0.11  
LGBMClassifier                       1.98  
XGBClassifier                        3.60  
BaggingClassifier                    0.28  
NearestCentroid                      0.08  
RandomForestClassifier               1.54  
BernoulliNB                          0.08  
LinearDiscriminantAnalysis           0.18  
ExtraTreesClassifier                 1.13  
GaussianNB                           0.05  
ExtraTreeClassifier                  0.06  
AdaBoostClassifier                   1.11  
PassiveAggressiveClassifier          0.17  
LogisticRegression                   1.11  
SGDClassifier                        0.40  
SVC                                  2.46  
Perceptron                           0.15  
CalibratedClassifierCV              13.71  
LinearSVC                            3.09  
QuadraticDiscriminantAnalysis        0.11  
LabelSpreading                      10.02  
LabelPropagation                     8.30  
RidgeClassifier                      0.11  
RidgeClassifierCV                    0.18  
KNeighborsClassifier                 1.29  
DummyClassifier                      0.05

- `Nenhum resample`

In [53]:
clf = LazyClassifier(verbose=0, predictions=True, ignore_warnings=True, custom_metric=None)
models_class, predictions_class = clf.fit(X_train, X_valid, y_train, y_valid)
models_class

100%|██████████| 29/29 [01:11<00:00,  2.47s/it]


Accuracy  Balanced Accuracy ROC AUC  F1 Score  \
Model                                                                          
XGBClassifier                      0.99               0.87    None      0.99   
LGBMClassifier                     0.99               0.87    None      0.99   
BaggingClassifier                  0.99               0.87    None      0.99   
DecisionTreeClassifier             0.99               0.84    None      0.99   
NearestCentroid                    0.75               0.77    None      0.78   
AdaBoostClassifier                 0.96               0.71    None      0.96   
RandomForestClassifier             0.98               0.69    None      0.98   
LinearDiscriminantAnalysis         0.87               0.61    None      0.86   
ExtraTreesClassifier               0.96               0.60    None      0.95   
BernoulliNB                        0.92               0.60    None      0.93   
PassiveAggressiveClassifier        0.86               0.58    None      0.87   
ExtraTreeClassifier                0.86               0.56    None      0.86   
GaussianNB                         0.49               0.54    None      0.57   
Perceptron                         0.84               0.51    None      0.85   
SGDClassifier                      0.89               0.50    None      0.88   
SVC                                0.92               0.50    None      0.91   
LogisticRegression                 0.89               0.49    None      0.88   
CalibratedClassifierCV             0.89               0.46    None      0.87   
LabelSpreading                     0.82               0.46    None      0.81   
LabelPropagation                   0.82               0.46    None      0.81   
LinearSVC                          0.89               0.46    None      0.87   
KNeighborsClassifier               0.84               0.42    None      0.83   
QuadraticDiscriminantAnalysis      0.49               0.42    None      0.57   
RidgeClassifier                    0.88               0.41    None      0.86   
RidgeClassifierCV                  0.88               0.41    None      0.86   
DummyClassifier                    0.59               0.24    None      0.59   

                               Time Taken  
Model                                      
XGBClassifier                        2.77  
LGBMClassifier                       2.12  
BaggingClassifier                    0.40  
DecisionTreeClassifier               0.13  
NearestCentroid                      0.08  
AdaBoostClassifier                   1.45  
RandomForestClassifier               1.66  
LinearDiscriminantAnalysis           0.21  
ExtraTreesClassifier                 1.46  
BernoulliNB                          0.11  
PassiveAggressiveClassifier          0.16  
ExtraTreeClassifier                  0.06  
GaussianNB                           0.05  
Perceptron                           0.17  
SGDClassifier                        0.84  
SVC                                  5.62  
LogisticRegression                   1.34  
CalibratedClassifierCV              17.16  
LabelSpreading                      15.45  
LabelPropagation                    11.29  
LinearSVC                            4.26  
KNeighborsClassifier                 3.72  
QuadraticDiscriminantAnalysis        0.20  
RidgeClassifier                      0.17  
RidgeClassifierCV                    0.44  
DummyClassifier                      0.05

# Análise sobre a função de custo

#### FUNÇÃO DE CUSTO COM CONFIGURAÇÃO ORIGINAL E SEM DESBALANCEAMENTO
##### Acurácia: 0.28  
##### Acurácia balanceada: 0.35  

           1       0.84      0.28      0.42      1484  
           2       0.37      0.28      0.32       509  
           3       0.03      0.34      0.05        47  
           4       0.01      0.50      0.01         8  
              precision    recall  f1-score   support
    
    accuracy                           0.28      2048  
    macro avg      0.31      0.35      0.20      2048  
    weighted avg   0.70      0.28      0.39      2048  


In [61]:
#                   sklearn - LogisticRegression     Acc   	B-Acc   F1 Score <- FUNÇÃO DE CUSTO <- ORIGINAL 
#                                                   0.31     0.70       0.39  
#                           ---- LazyPredict ----
# Sem resample              - LogisticRegression	0.89	 0.52	    0.89 <- FUNÇÃO DE CUSTO <- LazyPredict  
# Tomek                     - LogisticRegression	0.91	 0.51	    0.91  
# NeighbourhoodCleaningRule - LogisticRegression	0.99	 0.57	    0.99  

# Modelo com melhor desempenho - BaggingClassifier
## LazyClassifier com `NeighbourhoodCleaningRule resample` 
### X_remaining, y_remaining `não` foi `desbalanceado`

### Features utilizadas

- `features`  
['maximum_nights',
 'review_scores_cleanliness',
 'host_neighbourhood',
 'minimum_nights_avg_ntm',
 'host_response_rate',
 'accommodates',
 'neighbourhood_cleansed',
 'bedrooms',
 'number_of_reviews',
 'host_response_time',
 'review_scores_rating',
 'minimum_maximum_nights',
 'review_scores_location',
 'maximum_minimum_nights',
 'neighbourhood',
 'host_acceptance_rate',
 'description',
 'availability_365',
 'review_scores_checkin',
 'availability_60',
 'number_of_reviews_ltm',
 'first_review',
 'property_type',
 'number_of_reviews_l30d',
 'host_about',
 'neighborhood_overview',
 'host_location',
 'bathrooms_text',
 'host_name',
 'calculated_host_listings_count',
 'reviews_per_month',
 'host_listings_count',
 'last_review',
 'host_verifications',
 'host_since',
 'name',
 'maximum_nights_avg_ntm',
 'minimum_nights',
 'price',
 'beds']

In [57]:
clf_ncr = LazyClassifier(verbose=0, ignore_warnings=True, custom_metric=None)
# models_class_ncr, predictions_class_ncr = clf_ncr.fit(X_ncr_train, X_ncr_valid, y_ncr_train, y_ncr_valid)
final_models, final_prediction = clf_ncr.fit(X_ncr_train[features], X_remaining[features], y_ncr_train, y_remaining)
final_models

100%|██████████| 29/29 [01:03<00:00,  2.17s/it]


Accuracy  Balanced Accuracy ROC AUC  F1 Score  \
Model                                                                          
BaggingClassifier                  0.94               0.65    None      0.93   
XGBClassifier                      0.95               0.65    None      0.94   
LGBMClassifier                     0.95               0.60    None      0.94   
DecisionTreeClassifier             0.92               0.59    None      0.91   
NearestCentroid                    0.75               0.58    None      0.76   
GaussianNB                         0.70               0.55    None      0.72   
LinearDiscriminantAnalysis         0.77               0.47    None      0.72   
Perceptron                         0.78               0.47    None      0.73   
AdaBoostClassifier                 0.76               0.46    None      0.68   
RandomForestClassifier             0.88               0.45    None      0.86   
SGDClassifier                      0.80               0.44    None      0.75   
LogisticRegression                 0.77               0.43    None      0.72   
PassiveAggressiveClassifier        0.76               0.42    None      0.70   
BernoulliNB                        0.82               0.38    None      0.79   
LinearSVC                          0.75               0.38    None      0.67   
LabelSpreading                     0.75               0.37    None      0.67   
LabelPropagation                   0.75               0.37    None      0.67   
ExtraTreeClassifier                0.78               0.36    None      0.73   
KNeighborsClassifier               0.73               0.31    None      0.62   
CalibratedClassifierCV             0.74               0.30    None      0.65   
QuadraticDiscriminantAnalysis      0.76               0.30    None      0.71   
SVC                                0.74               0.30    None      0.64   
ExtraTreesClassifier               0.73               0.29    None      0.63   
RidgeClassifierCV                  0.72               0.25    None      0.61   
RidgeClassifier                    0.72               0.25    None      0.61   
DummyClassifier                    0.72               0.25    None      0.61   

                               Time Taken  
Model                                      
BaggingClassifier                    0.27  
XGBClassifier                        2.05  
LGBMClassifier                       2.16  
DecisionTreeClassifier               0.12  
NearestCentroid                      0.18  
GaussianNB                           0.11  
LinearDiscriminantAnalysis           0.21  
Perceptron                           0.18  
AdaBoostClassifier                   1.31  
RandomForestClassifier               1.33  
SGDClassifier                        0.33  
LogisticRegression                   1.06  
PassiveAggressiveClassifier          0.16  
BernoulliNB                          0.15  
LinearSVC                            0.77  
LabelSpreading                      18.31  
LabelPropagation                    16.69  
ExtraTreeClassifier                  0.10  
KNeighborsClassifier                11.66  
CalibratedClassifierCV               2.58  
QuadraticDiscriminantAnalysis        0.25  
SVC                                  1.18  
ExtraTreesClassifier                 1.12  
RidgeClassifierCV                    0.31  
RidgeClassifier                      0.22  
DummyClassifier                      0.09